src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -447,6 +448,15 @@ InterpreterSelectQuery::InterpreterSelectQuery(
 
     auto analyze = [&] (bool try_move_to_prewhere)
     {
+        if (context->hasQueryContext())
+        {
+            size_t & current_query_analyze_count = context->getQueryContext()->kitchen_sink.analyze_counter;
+            current_query_analyze_count++;
+            if (current_query_analyze_count >= settings.max_pipeline_depth)
+                throw DB::Exception(ErrorCodes::TOO_DEEP_PIPELINE, "Query analyze overflow. Try to increase `max_pipeline_depth` or simplify the query" );
```

Comment:
Also, we planned to use this counter to inhibit the second analysis step.

Reply:
In which cases second analysis step should be inhibited?

Reply:
It was said that the second analysis step is only an optimization. (Although I'm not sure)
If we already have made too many analyses, we can avoid further optimizations.

Also you can remove exponential blowup by memorizing the ASTs that were already optimized.

Reply:
Changed it to use `max_pipeline_depth / 10` as a the threshold to disable the second analyze and `max_pipeline_depth` to cancel the query at all.